iroh-relay/src/quic.rs

```diff
@@ -0,0 +1,301 @@
+                        if let Err(err) = res {
+                            if err.is_panic() {
+                                panic!("task panicked: {:#?}", err);
+                            }
+                            warn!("connection failed: {err}");
+                        }
+                    }
+                    res = endpoint.accept() => match res {
+                        Some(conn) => {
+                             debug!("accepting connection from {:?}", conn.remote_address())       ;
+                             set.spawn(async move {
+                                 let remote_addr = conn.remote_address();
+                                 let res = handle_connection(conn).await;
+                                 if let Err(ref err) = res {
+                                     warn!(remote_address = ?remote_addr, "error handling connection {err:?}")
+                                 }
+                                 res
+                             });
+                        }
+                        None => {
+                            debug!("endpoint closed");
+                            break;
+                        }
+                    }
+                }
+            }
+            endpoint
+                .close(QUIC_ADDR_DISC_CLOSE_CODE, QUIC_ADDR_DISC_CLOSE_REASON);
+            endpoint.wait_idle().await;
+            debug!("quic endpoint has been shutdown.");
+        }.instrument(info_span!("quic-endpoint")),);
+            Ok(Self {
+                bind_addr,
+                cancel,
+                handle: AbortOnDropHandle::new(task),
+            })
+        }
+
+        /// Closes the underlying QUIC endpoint and the tasks running the
+        /// QUIC connections.
+        pub fn shutdown(&self) {
+            self.cancel.cancel();
+        }
+    }
+
+    /// A handle for the Server side of QUIC address discovery.
+    ///
+    /// This does not allow access to the task but can communicate with it.
+    #[derive(Debug, Clone)]
+    pub struct ServerHandle {
+        cancel_token: CancellationToken,
+    }
+
+    impl ServerHandle {
+        /// Gracefully shut down the quic endpoint.
+        pub fn shutdown(&self) {
+            self.cancel_token.cancel()
+        }
+    }
+
+    async fn handle_connection(conn: quinn::Incoming) -> Result<()> {
+        let connection = conn.await?;
+        info!("established");
+        // wait for the client to close the connection
+        let connection_err = connection.closed().await;
+        match connection_err {
+            quinn::ConnectionError::ApplicationClosed(ApplicationClose { error_code, .. })
+                if error_code == QUIC_ADDR_DISC_CLOSE_CODE =>
+            {
+                return Ok(());
+            }
+            _ => {
+                warn!(
+                    "{} - error closing connection {connection_err:?}",
+                    connection.remote_address()
```

Comment:
```suggestion
                    error_code, reason, "error closing connection {connection_err:#}"
```
The address is already part of the span.

You'll need to expose reason to add it as a field.

Reply:
not all connection errors have error_codes or reasons, but I see what you are saying!